src/app/server/Server.h

```diff
@@ -386,13 +386,15 @@ class Server
         return System::SystemClock().GetMonotonicMicroseconds64() - mInitTimestamp;
     }
 
-    static Server & GetInstance() { return sServer; }
+    static Server & GetInstance()
+    {
+        static Server * sServer = new Server();
```

Comment:
do you need to use heap here? Could this be `static server sServer; return &sServer` ?

Reply:
There are potential issues with that approach regarding deinitialization: https://isocpp.org/wiki/faq/ctors#construct-on-first-use-v2

I don't feel qualified to say that they're not a problem here, but if you tell me they aren't a problem then I can change it :)

Reply:
I don't think this is an initialization order problem as such... The problem is that `PairingSession` calls `GetLocalMRPConfig` during construction, rather than just before we actually start using it.

Reply:
Ah yeah if we can move the GetLocalMRPConfig call out of the constructor that would also solve this.

Reply:
So the issue here is that when `CONFIG_BUILD_FOR_HOST_UNIT_TEST` is set the static constructor for `Server` can run before the static constructor for `idleRetransTimeoutOverride` and whatnot, and then we get use-before-initialization?

Realistically, calling GetLocalMRPConfig "too early" is just broken, because the whole point of not just those test overrides but SetLocalMRPConfig is that it can change.  And in practice, the subclasses of PairingSession do the following:

1. `CASESession::EstablishSession` (CASE initiator init) sets `mLocalMRPConfig`.
2. `CASESession::PrepareForSessionEstablishment` (CASE responder init) sets `mLocalMRPConfig`.
3. `PASESession::WaitForPairing` (PASE responder init) sets `mLocalMRPConfig`.
4. `PASESession::Pair` (PASE initiator init) sets `mLocalMRPConfig`.

So it really does not matter what `mLocalMRPConfig` is set to in the PairingSession constructor.  I would say we should just switch it to either default-initialization, of if that does not work because there is no default ctor just set it to GetDefaultMRPConfig(), with a comment explaining why it shouldn't use GetLocalMRPConfig() and that the value does not matter because actual pairing session instances always override it.

@mwswartwout thoughts?

Reply:
Or get rid of `mLocalMRPConfig` and just call `GetLocalMRPConfig()` inside `PairingSession::EncodeSessionParameters`?
The current behavior may be a bit counterintuitive that MRP config can be overriden but `PairingSession` may still use some cached value from the past (e.g. from when `PrepareForSessionEstablishment` was called). The main users of setting `mLocalMRPConfig`  upfront are probably unit tests but they could probably also set `idleRetransTimeoutOverride` if needed.

Reply:
@bzbarsky-apple @ksperling-apple @Damian-Nordic thoughts on making `mLocalMRPConfig` an `Optional` vs removing it entirely?

Reply:
Sorry for the lag.  We could make it an Optional, but that has some storage cost (one extra word), right?   The benefit compared to what I suggested is that it would let us add those asserts, right?

I don't see how we can remove it without changing the API promise, that being that you can provide an override value when you start session establishment, right?  And maybe that's OK; there keeps being this tension between "inject everything so we don't have global functions" and "just get the thing when you need it", and if local MRP config is dynamically changeable (as it is) just getting it at the point when you need it makes the most sense.

Reply:
I don't see how removing `mLocalMRPConfig` changes the API promise of being able to provide override values. In fact what it does is allow you to provide the override later and still have it take effect. But if you continue to set the overrides before you start session establishment then the overrides will still be respected when they are fetched later.

Reply:
Right now you are able to provide an override that is _not the same as what GetLocalMRPConfig() returns_.

Reply:
> Where are those overrides stored

They're arguments to `CASESession::PrepareForSessionEstablishment`, `CASESession::EstablishSession`, and equivalent PASESession APIs.  If you're using those classes directly, you can just pass them in.

If you're using `CASESession` via `CASESessionManager`, say, then `CASESessionManager` can be initialized with a `CASESessionManagerConfig` which has `CASEClientInitParams` which allows passing in a `mrpLocalConfig`.  Generally speaking people shouldn't, but they can....

Now maybe we should in fact remove this API surface.  I'd be ok with that personally, but it might need some wider discussion.

Reply:
Ah okay yeah I got it now. Let me update mLocalMRPConfig to be an Optional and see what kind of size hit we take, and then removing it entirely can be a separate conversation.